inspirehep/modules/orcid/api.py

```diff
@@ -63,12 +70,19 @@ def push_record_with_orcid(recid, orcid, oauth_token, put_code=None, old_hash=No
     """
     record = get_db_record('lit', recid)
 
-    new_hash = calculate_hash_for_record(record)
+    orcid_cache = OrcidCache(orcid)
+
+    if not putcode or not old_hash:
+        cached_putcode, cached_old_hash = orcid_cache.read_record_data(recid)
+        putcode = putcode or cached_putcode
+        old_hash = old_hash or cached_old_hash
+
+    new_hash = OrcidHasher().compute_hash(record)
```

Comment:
I think I would make the hashing part of the `OrcidCache`, which you initialize with `orcid` and `record`, and acts like an oracle that tells you whether to `post`, `put` or do nothing.

Reply:
I like your idea of hiding the hasher inside the OrcidCache either by encapsulation or association. This way the code that uses OrcidCache only knows about the concept of caching and not hashing. I will try this.